thirdparty/faiss/faiss/IndexIVF.h

```diff
@@ -503,6 +503,28 @@ struct InvertedListScanner {
             size_t k,
             size_t& scan_cnt) const;
 
+    /** scan a set of codes, compute distances to current query and
+     * push all to heap. Default implemetation
+     * calls distance_to_code.
+     *
+     * @param list_size     number of codes to scan
+     * @param codes         codes to scan (list_size * code_size)
+     * @param code_norms    norms of code (for cosine)
+     * @param ids           corresponding ids (ignored if store_pairs)
+     * @param distances     heap distances (size counter_back)
+     * @param labels        heap labels (size counter_back)
+     * @param counter_back  heap size (will increase)
+     * @return number of heap pushes performed
+     */
+    virtual size_t scan_codes_and_push_back(
```

Comment:
Why is this function needed at all? Why can't a plain `scan_codes()` be used with a heap large enough to hold all needed items?

Reply:
The main reason lies in the different orientations of the heaps used by `ann-search` and `iterator-search`.
Taking `L2` as an example, the iterator requires a `minheap` to ensure that each `next()` operation returns the closest one.
On the other hand, the ann-search aims to find the k-nearest nodes, so it uses a `maxheap` and reorders the results. Obviously, complete sorting is unnecessary for the iterator.